### PR TITLE
[DataFrame] Implement quantile

### PR DESCRIPTION
**Pull request by 11rohans**
Updates the `DataFrame.quantile` method for full use cases. 
Also updates the init file to account for datetime methods. 
Updates equals method so that it can handle one column dataframes correctly

### CODE REVIEW

File: python/ray/dataframe/dataframe.py [hunk 1]
```diff
@@ -2774,37 +2773,66 @@ def quantile(self, q=0.5, axis=0, numeric_only=True,
...
-                quantiles.append(result)
-
-            return pd.concat(quantiles, axis=1).T
+            if not all(0 <= qs <= 1 for qs in q):
+                raise ValueError(msg.format(np.asarray(q) / 100.0))
         else:
+            if not (0 <= q <= 1):
+                raise ValueError(msg.format(q / 100.0))
+
+        if isinstance(q, (pd.Series, np.ndarray, pd.Index, list)):
```
Comment (kunalgosar) [inline python/ray/dataframe/dataframe.py]: Use `axis = pd.DataFrame()._get_axis_number(axis)` instead.

File: python/ray/dataframe/dataframe.py [hunk 2]
```diff
@@ -2774,37 +2773,66 @@ def quantile(self, q=0.5, axis=0, numeric_only=True,
...
         """
 
-        def quantile_helper(df, q, axis, numeric_only, interpolation):
-            try:
-                return df.quantile(q=q, axis=axis, numeric_only=numeric_only,
-                                   interpolation=interpolation)
-            except ValueError:
-                return pd.Series()
+        # deal with error checking for bad dtyping when numeric_only is False
+        if not numeric_only:
```
Comment (kunalgosar) [inline python/ray/dataframe/dataframe.py]: `numeric_only` specifies whether or not to filter the non-numeric columns. Does not need error checking.

File: python/ray/dataframe/dataframe.py [hunk 3]
```diff
@@ -2774,37 +2773,66 @@ def quantile(self, q=0.5, axis=0, numeric_only=True,
...
+            # check if there are any object columns
+            for t in self.dtypes:
+                if t == np.dtype('O'):
+                    e_msg = "Cannot call quantile on columns of type " + str(t)
+                    raise TypeError(e_msg)
+
+        # check that all qs are between 0 and 1
+        msg = ("percentiles should all be in the interval [0, 1]. "
```
Comment (kunalgosar) [inline python/ray/dataframe/dataframe.py]: Use the pandas `_check_percentile` method here.

File: python/ray/dataframe/dataframe.py [hunk 4]
```diff
@@ -2774,37 +2773,66 @@ def quantile(self, q=0.5, axis=0, numeric_only=True,
...
+                                 columns=self.index)
+
+        # single q means we can use arithmetic helper
+        else:
+            def quantile_helper(df, q, axis, numeric_only, interpolation):
+                try:
+                    return df.quantile(q=q, axis=axis,
+                                       numeric_only=numeric_only,
+                                       interpolation=interpolation)
+                except ValueError:
```
Comment (kunalgosar) [inline python/ray/dataframe/dataframe.py]: When would this ValueError be thrown?

File: python/ray/dataframe/dataframe.py [hunk 5]
```diff
@@ -2774,37 +2773,66 @@ def quantile(self, q=0.5, axis=0, numeric_only=True,
...
+
             def remote_func(df):
                 return quantile_helper(df, q=q, axis=axis,
                                        numeric_only=numeric_only,
                                        interpolation=interpolation)
-
             result = self._arithmetic_helper(remote_func, axis)
             result.name = q
-            return result
+            return pd.Series(result)
```
Comment (kunalgosar) [inline python/ray/dataframe/dataframe.py]: `_arithmetic_helper` should return a Series.

**Review by kunalgosar — CHANGES_REQUESTED** [review]
Looks really good, just a few comments. I wasn't able to reproduce a `ValueError` from the underlying pandas call. Can you describe when this would occur?

File: python/ray/dataframe/dataframe.py (hunk 4 above)
Comment (11rohans) [reply python/ray/dataframe/dataframe.py]: Occurs when a dataframe partition has mixed dtypes

Comment (kunalgosar) [reply python/ray/dataframe/dataframe.py]: I see, this should probably be handled somewhere to ensure concordance.

File: python/ray/dataframe/dataframe.py (hunk 2 above)
Comment (11rohans) [reply python/ray/dataframe/dataframe.py]: If numeric_only is true, then pandas returns value errors for mixed dtype dataframes

File: python/ray/dataframe/dataframe.py (hunk 3 above)
Comment (11rohans) [reply python/ray/dataframe/dataframe.py]: Updated

File: python/ray/dataframe/dataframe.py (hunk 1 above)
Comment (11rohans) [reply python/ray/dataframe/dataframe.py]: What would this do here? I'm checking if q is a list or not

File: python/ray/dataframe/dataframe.py (hunk 2 above)
Comment (kunalgosar) [reply python/ray/dataframe/dataframe.py]: I'm not able to reproduce, seems that pandas drops non-numeric columns when `numeric_only` is True.

```
In [3]: df
Out[3]: 
   col1 col2
0     1    a
1     2    b
2     3    c

In [4]: df.dtypes
Out[4]: 
col1     int64
col2    object
dtype: object

In [5]: df.quantile(numeric_only=True)
Out[5]: 
col1    2.0
Name: 0.5, dtype: float64
```

Comment (kunalgosar) [reply python/ray/dataframe/dataframe.py]: This is true also in the `axis=1` case.
```
In [6]: df.quantile(numeric_only=True, axis=1)
Out[6]: 
0    1.0
1    2.0
2    3.0
Name: 0.5, dtype: float64
```

File: python/ray/dataframe/dataframe.py (hunk 4 above)
Comment (11rohans) [reply python/ray/dataframe/dataframe.py]: Describe also handles it like this

File: python/ray/dataframe/dataframe.py (hunk 2 above)
Comment (11rohans) [reply python/ray/dataframe/dataframe.py]: Whoops, when numeric_only is FALSE, it returns TYPEerrors. 

Try on 

df = pd.DataFrame({"A": [1, 2, 3],
            "B": [2., 3., 4.],
            "C": pd.date_range('20130101', periods=3),
            "D": ['foo', 'bar', 'baz']})
df.quantile(.5, axis=1, numeric_only=False)

Comment (kunalgosar) [reply python/ray/dataframe/dataframe.py]: This is true, although the check here does not handle this case. The error I see is `TypeError: Cannot compare type 'Timestamp' with type 'float'`.

File: python/ray/dataframe/dataframe.py [hunk 6]
```diff
@@ -2849,34 +2848,57 @@ def quantile(self, q=0.5, axis=0, numeric_only=True,
...
+        # single q means we can use arithmetic helper
         else:
+            def quantile_helper(df, q, axis, numeric_only, interpolation):
+                try:
+                    return df.quantile(q=q, axis=axis,
+                                       numeric_only=numeric_only,
+                                       interpolation=interpolation)
+                # This exception is thrown when there are only non-numeric columns
+                # in this partition
+                except ValueError:
```
Comment (kunalgosar) [inline python/ray/dataframe/dataframe.py]: As discussed above, `TypeErrors` are thrown here. Perhaps you can add some logic at the beginning of the function to ensure all the `dtypes` are comparable. 

The issue seems to arise not from non-numeric data, but from non-comparable types.

**Review by kunalgosar — CHANGES_REQUESTED** [review]
Looks much better; should probably look into the right error handling here. Seems like ignoring errors on workers is the wrong solution.

It would be better if the errors were caught in the function logic, before submitting remote tasks. Since this is an issue in `describe` can you update there too?

File: python/ray/dataframe/__init__.py [hunk 7]
```diff
@@ -3,7 +3,9 @@
 from __future__ import print_function
 
 import pandas as pd
-from pandas import (eval, unique, value_counts, Panel, date_range, MultiIndex)
+from pandas import (eval, MultiIndex, CategoricalIndex, Series, Panel,
+                    unique, value_counts,
+                    date_range, Timedelta, Timestamp, to_datetime, NaT)
```
Comment (kunalgosar) [inline python/ray/dataframe/__init__.py]: The lines should be consolidated.

File: python/ray/dataframe/__init__.py [hunk 8]
```diff
@@ -29,7 +31,6 @@ def get_npartitions():
 # We import these file after above two function
 # because they depend on npartitions.
 from .dataframe import DataFrame  # noqa: 402
-from .series import Series  # noqa: 402
```
Comment (kunalgosar) [inline python/ray/dataframe/__init__.py]: If you do this, you need to remove `test_series`.

**Review by kunalgosar — CHANGES_REQUESTED** [review]
Few preliminary comments.

File: python/ray/dataframe/dataframe.py [hunk 9]
```diff
@@ -2882,35 +2881,72 @@ def quantile(self, q=0.5, axis=0, numeric_only=True,
...
+            # check if there are any object columns
+            for t in self.dtypes:
+                if t == np.dtype('O'):
+                    e_msg = "Cannot call quantile on columns of type " + str(t)
+                    raise TypeError(e_msg)
+        else:
+            if all([dtype == np.dtype('O') or
+                    np.issubdtype(dtype, np.datetime64)
+                    for dtype in self.dtypes]):
+                raise ValueError("need at least one array to concatenate")
```
Comment (kunalgosar) [inline python/ray/dataframe/dataframe.py]: I know that this error is concordant with Pandas, but it feels very out of place in this context. Perhaps we should change it to something more descriptive. Honestly, this case seems like a bug with Pandas, it should handle the case where all columns are dropped.

Looping in @devin-petersohn here.

File: python/ray/dataframe/dataframe.py [hunk 10]
```diff
@@ -2882,35 +2881,72 @@ def quantile(self, q=0.5, axis=0, numeric_only=True,
...
+                        all([(dtype == np.dtype('O') or
+                              np.issubdtype(dtype, np.datetime64))
+                            for dtype in df.dtypes])):
+                    return pd.DataFrame()
+                else:
+                    return df.quantile(q=q, axis=axis,
+                                       numeric_only=numeric_only,
+                                       interpolation=interpolation)
+
+            if axis == 0 or axis == 'index':
```
Comment (kunalgosar) [inline python/ray/dataframe/dataframe.py]: As before use `axis = pd.DataFrame()._get_axis_number(axis)` here

File: python/ray/dataframe/dataframe.py [hunk 11]
```diff
@@ -2882,35 +2881,72 @@ def quantile(self, q=0.5, axis=0, numeric_only=True,
...
+        # deal with error checking for bad dtyping when numeric_only is False
+        if not numeric_only:
+            # check if there are any object columns
+            for t in self.dtypes:
+                if t == np.dtype('O'):
+                    e_msg = "Cannot call quantile on columns of type " + str(t)
+                    raise TypeError(e_msg)
+        else:
+            if all([dtype == np.dtype('O') or
+                    np.issubdtype(dtype, np.datetime64)
```
Comment (kunalgosar) [inline python/ray/dataframe/dataframe.py]: Seems that `np.datetime64` type columns are supported.

```
In [8]: x
Out[8]: 
                 col1
0 2018-05-04 20:50:29
1 2018-05-04 20:50:29
2 2018-05-04 20:50:29

In [9]: x.dtypes
Out[9]: 
col1    datetime64[ns]
dtype: object

In [10]: x.quantile(q=0.5, axis=1, numeric_only=True)
Out[10]: 
0   NaN
1   NaN
2   NaN
Name: 0.5, dtype: float64
```

File: python/ray/dataframe/dataframe.py [hunk 12]
```diff
@@ -2882,35 +2881,72 @@ def quantile(self, q=0.5, axis=0, numeric_only=True,
...
+                return DataFrame(col_partitions=result,
+                                 index=q_index,
+                                 columns=self.index)
+
+        # single q means we can use arithmetic helper
         else:
-            def remote_func(df):
-                return quantile_helper(df, q=q, axis=axis,
+
+            def quantile_helper(df):
```
Comment (kunalgosar) [inline python/ray/dataframe/dataframe.py]: Why is this function defined twice?

**Review by kunalgosar — CHANGES_REQUESTED** [review]
A few comments. Looping in @devin-petersohn to discuss the error message shown.

File: python/ray/dataframe/__init__.py (hunk 7 above)
Comment (11rohans) [reply python/ray/dataframe/__init__.py]: Wanted to keep all the datetime imports on their own line, should I still do this?

File: python/ray/dataframe/__init__.py (hunk 8 above)
Comment (11rohans) [reply python/ray/dataframe/__init__.py]: Should I just move the entire file?

File: python/ray/dataframe/dataframe.py (hunk 9 above)
Comment (11rohans) [reply python/ray/dataframe/dataframe.py]: I agree -- I originally had a more descriptive error but scrapped it in favor of using the pandas one. Something that references mixed dtypes would be better

File: python/ray/dataframe/dataframe.py (hunk 12 above)
Comment (11rohans) [reply python/ray/dataframe/dataframe.py]: One returns a Series the other returns a DataFrame

File: python/ray/dataframe/dataframe.py (hunk 11 above)
Comment (11rohans) [reply python/ray/dataframe/dataframe.py]: Made a fix to handle all datetime objects

File: python/ray/dataframe/__init__.py (hunk 8 above)
Comment (robertnishihara) [reply python/ray/dataframe/__init__.py]: Don't we want to run those tests? Or no?

Comment (kunalgosar) [reply python/ray/dataframe/__init__.py]: While Series is unimplemented, we are importing it directly from Pandas. Skipping these tests for now as they expect `NotImplementedError`s to be thrown.

**Review by devin-petersohn — APPROVED** [review]

**Comment by devin-petersohn** [comment]
Merged, thanks @11rohans!

**Pull request merged**